tests/cephadm/test_custom_container.py

```diff
@@ -0,0 +1,157 @@
+    """
+
+    # Get installer node
+    installer = ceph_cluster.get_nodes(role="installer")[0]
+
+    # Get config file and node
+    config = kw.get("config")
+
+    # Get cluster nodes
+    ceph_nodes = kw.get("ceph_nodes")
+
+    # Get dashboard image
+    registry_image = config.get("dashboard_images")
+    dashboard_images = registry_image.values()
+
+    # Get registry name
+    registry = [reg.split("/")[0] for reg in dashboard_images][0]
+
+    # Get registry docker image
+    docker_reg_image = config.get("registry_image").get("docker_registry_image")
+
+    # Get the rhbuild
+    rhbuild = config.get("rhbuild")
+    if rhbuild.startswith("6"):
+        rhbuild = "quincy"
+    elif rhbuild.startswith("5"):
+        rhbuild = "pacific"
+    else:
+        raise Exception("Unsupported version")
+
+    # Setup ssh key to all nodes
+    setup_ssh_keys(installer, ceph_nodes)
+
+    # Get registry credentials
+    con = get_cephci_config()
+    reg_credentials = con.get("cdn_credentials", {})
+    reg_username = reg_credentials.get("username")
+    reg_password = reg_credentials.get("password")
+
+    # Setup a private registry on installer node
+    setup_private_container_registry(
+        installer=installer,
+        nodes=ceph_nodes,
+        registry=registry,
+        reg_username=reg_username,
+        reg_password=reg_password,
+        build_type=rhbuild,
+        private_reg_username=PRIVATE_REG_USER,
+        private_reg_password=PRIVATE_REG_PASS,
+        detach=docker_reg_image,
+        images=dashboard_images,
+    )
+
+    # Get private registry catalog
+    out = get_private_registry_image(
+        installer, username=PRIVATE_REG_USER, password=PRIVATE_REG_PASS
+    )
+    repos = out.get("repositories")
+
+    # Pull images to nodes
+    for ceph in ceph_nodes:
+        Registry(ceph).login(
+            registry=f"{installer.hostname}:{PORT}",
+            username=PRIVATE_REG_USER,
+            password=PRIVATE_REG_PASS,
+        )
+        for img in repos:
+            Container(ceph).pull(image=f"{installer.hostname}:{PORT}/{img}")
+
+    # Define dashboard stack variable
+    services = ["dashboard", "alertmanager", "node-exporter", "prometheus"]
+    service_config = {
+        "prometheus": "container_image_prometheus",
+        "dashboard": "container_image_grafana",
+        "alertmanager": "container_image_alertmanager",
+        "node-exporter": "container_image_node_exporter",
+    }
+
+    # Set custom image to dashboard stack
+    for service in services:
+        if service == "prometheus":
+            continue
+
+        config_key = service_config.get(service)
+        if config_key:
+            for repo in repos:
+                if service in repo:
+                    conf = {
+                        "key": "mgr",
+                        "value": f"mgr/cephadm/{config_key} {installer.hostname}:{PORT}/{repo}",
+                    }
+                    CephAdm(installer).ceph.config.set(**conf)
```

Comment:
`CephAdm` class object is used multiple times, better to create this object and reuse it.